### PR TITLE
Replace simple_html_dom with Symfony\DomCrawler

### LINKED ISSUE #434: Replace simple_html_dom
Replace with `symfony/dom-crawler`

### PR DESCRIPTION
**Pull request by dvesh3**
## Changes in this pull request  
Resolves #434

### CODE REVIEW

File: bundles/AdminBundle/Controller/Admin/TranslationController.php [hunk 1]
```diff
@@ -1033,34 +1030,26 @@ public function wordExportAction(Request $request)
...
 
                         // remove empty tags (where it matters)
-                        $elements = $dom->find('a, li');
-                        if ($elements) {
-                            foreach ($elements as $el) {
-                                $string = $clearText($el->plaintext);
-                                if (empty($string)) {
-                                    $el->outertext = '';
-                                }
+                        // replace links => links get [Linktext]
```
Comment (brusch) [inline bundles/AdminBundle/Controller/Admin/TranslationController.php:1033]: `li` tags were covered here as well in the past, or is it not needed anymore?

File: lib/Http/Response/CodeInjector.php [hunk 2]
```diff
@@ -118,20 +119,20 @@ private function injectIntoPresetSelector(string $html, string $code, string $se
 
     private function injectIntoDomSelector(string $html, string $code, string $selector, string $position, string $charset): string
     {
-        $dom = str_get_html($html);
+        $dom = new Crawler($html);
         if ($dom) {
```
Comment (brusch) [inline lib/Http/Response/CodeInjector.php:123]: This `if` can go - if there is something wrong with the HTML contents, `new Crawler()` would throw an exception.

File: lib/Mail.php [hunk 3]
```diff
@@ -760,15 +761,16 @@ public function getBodyTextRendered()
             //creating text version from html email
             try {
                 $htmlContent = $this->getBodyHtmlRendered();
-                $html = str_get_html($htmlContent);
+                $html = new Crawler($htmlContent);
+
                 if ($html) {
```
Comment (brusch) [inline lib/Mail.php:766]: Same here, this `if` can go.

File: lib/Templating/Renderer/IncludeRenderer.php [hunk 4]
```diff
@@ -171,16 +172,15 @@ protected function modifyEditmodeContent(PageSnippet $include, $content)
...
         // this is needed by the editmode to highlight included documents
-        if ($html = str_get_html($content)) {
-            $childs = $html->find('*');
-            if (is_array($childs)) {
-                foreach ($childs as $child) {
-                    $child->class = $child->class . $editmodeClass;
-                    $child->pimcore_type = $include->getType();
-                    $child->pimcore_id = $include->getId();
-                }
+        if ($html = new Crawler($content)) {
```
Comment (brusch) [inline lib/Templating/Renderer/IncludeRenderer.php:175]: Again the `if` 😊

File: lib/Tool/Newsletter.php [hunk 5]
```diff
@@ -80,16 +81,17 @@ public static function prepareMail(
 
         // render the document and rewrite the links (if analytics is enabled)
         if ($contentHTML && $newsletterDocument->getEnableTrackingParameters()) {
-            $html = str_get_html($contentHTML);
+            $html = new Crawler($contentHTML);
             if ($html) {
```
Comment (brusch) [inline lib/Tool/Newsletter.php:85]: `if` 😊

File: lib/Tool/Text.php [hunk 6]
```diff
@@ -186,31 +187,32 @@ public static function wysiwygText($text, $params = [])
     public static function replaceWysiwygTextRelationIds($idMapping, $text)
     {
         if (!empty($text)) {
-            $html = str_get_html($text);
+            $html = new Crawler($text);
             if (!$html) {
```
Comment (brusch) [inline lib/Tool/Text.php:191]: `if`, but maybe we need a `try/catch` here.

File: lib/Translation/Escaper/Xliff12Escaper.php [hunk 7]
```diff
@@ -92,15 +94,16 @@ public function unescapeXliff(string $content): string
         $content = $this->parseInnerXml($content);
 
         if (preg_match("/<\/?(bpt|ept)/", $content)) {
-            $xml = str_get_html($content);
+            $xml = new Crawler($content);
             if ($xml) {
```
Comment (brusch) [inline lib/Translation/Escaper/Xliff12Escaper.php:98]: `if` can go.

File: models/DataObject/ClassDefinition/Data/Wysiwyg.php [hunk 8]
```diff
@@ -361,31 +362,28 @@ public function getDiffVersionPreview($data, $object = null, $params = [])
...
      *
-     * @return Element\ElementInterface
+     * @return mixed
      */
     public function rewriteIds($object, $idMapping, $params = [])
     {
         $data = $this->getDataFromObjectParam($object, $params);
-        $html = str_get_html($data);
+        $html = new Crawler($data);
         if ($html) {
```
Comment (brusch) [inline models/DataObject/ClassDefinition/Data/Wysiwyg.php:371]: `if`

File: models/Document/Editable/Wysiwyg.php [hunk 9]
```diff
@@ -158,39 +159,32 @@ public function getCacheTags($ownerDocument, $blockedTags = [])
...
     public function rewriteIds($idMapping)
     {
-        $html = str_get_html($this->text);
-        if (!$html) {
-            return $this->text;
-        }
+        $html = new Crawler($this->text);
 
-        $s = $html->find('a[pimcore_id],img[pimcore_id]');
+        if ($html) {
```
Comment (brusch) [inline models/Document/Editable/Wysiwyg.php:169]: `if`

**Review by brusch — CHANGES_REQUESTED** [review]
I think it would make sense so include `masterminds/html5` in composer.json, details see: https://symfony.com/blog/new-in-symfony-4-3-better-html5-parser-for-domcrawler

File: bundles/AdminBundle/Controller/Admin/TranslationController.php (hunk 1 above)
Comment (dvesh3) [reply bundles/AdminBundle/Controller/Admin/TranslationController.php]: It will update anchors inside `li` tags too, so no need to filter by `li`

**Comment by brusch** [comment]
@dvesh3 thanks for the update, however it seems that the following is still missing: 
> I think it would make sense so include masterminds/html5 in composer.json, details see: https://symfony.com/blog/new-in-symfony-4-3-better-html5-parser-for-domcrawler

**Comment by dvesh3** [comment]
> @dvesh3 thanks for the update, however it seems that the following is still missing:
> 
> > I think it would make sense so include masterminds/html5 in composer.json, details see: https://symfony.com/blog/new-in-symfony-4-3-better-html5-parser-for-domcrawler

sorry, missed it in last commit. push it now 😉

**Pull request merged**